### PR TITLE
Derive a canonical URL from a page URL when a canonical tag is not present

### PR DESCRIPTION
**Pull request by teiswit**
Pages missing `<link rel="canonical">` could cause an infinite EA grant as we treat them as having the same canonical URL (i.e. an empty string). As canonical pages do not necessarily require the canonical tag, we should treat the page URL as canonical when the canonical tag is not present to avoid the infinite grant issue. 

Internal reference: b/256975824

### CODE REVIEW

File: src/utils/url.js [hunk 1]
```diff
@@ -199,12 +199,17 @@ export function getHostUrl(url) {
 }
 
 /**
+ * Returns the canonical URL from the canonical tag. If the canonical tag is
+ * not present, treat the doc URL itself as canonical.
  * @param {!../model/doc.Doc} doc
  * @return {string}
  */
 export function getCanonicalUrl(doc) {
   const node = doc.getRootNode().querySelector("link[rel='canonical']");
```
Comment (ChrisAntaki) [inline src/utils/url.js:208]: Looks like we call `doc.getRootNode()` 3x in this method. Would it make sense to call it once? Ex: `const rootNode = doc.getRootNode()`

File: src/utils/url.js [hunk 2]
```diff
@@ -199,12 +199,17 @@ export function getHostUrl(url) {
...
-  return (node && node.href) || '';
+  return (
+    (node && node.href) ||
```
Comment (ChrisAntaki) [inline src/utils/url.js:210]:
```suggestion
    node?.href ||
```

File: src/runtime/entitlements-manager-test.js [hunk 3]
```diff
@@ -103,6 +108,15 @@ describes.realWin('EntitlementsManager', {}, (env) => {
     deps = new DepsDef();
     sandbox.stub(deps, 'win').returns(win);
     const globalDoc = new GlobalDoc(win);
+
+    globalDoc.getRootNode = function () {
```
Comment (ChrisAntaki) [inline src/runtime/entitlements-manager-test.js:112]: Arrow functions could help this code be more concise. Ex:
```

**Comment by ChrisAntaki** [comment]
This LGTM overall

@jpettitt , could you verify this approach of falling back to the doc's origin+path (without query string) is alright?

File: src/runtime/entitlements-manager-test.js (hunk 3 above)
Comment (teiswit) [reply src/runtime/entitlements-manager-test.js:112]: Thanks for the suggestion. Simplified accordingly.

File: src/utils/url.js (hunk 1 above)
Comment (teiswit) [reply src/utils/url.js:208]: Totally makes sense. Revised accordingly.

**Review by ChrisAntaki — APPROVED** [review]

**Comment by ChrisAntaki** [comment]
Looks good! Thanks for being open to feedback

**Pull request merged**